client/grafana-data-source/src/database.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use crate::Error;
+
+pub struct Database {
+	base_timestamp: i64,
+	storage: HashMap<&'static str, Vec<Datapoint>>
+}
+
+impl Database {
+	pub fn new() -> Self {
+		Self {
+			base_timestamp: now_millis(),
+			storage: HashMap::new()
+		}
+	}
+
+	pub fn keys_starting_with<'a>(&'a self, base: &'a str) -> impl Iterator<Item = &'static str> + 'a {
+		self.storage.keys().filter(move |key| key.starts_with(base)).cloned()
+	}
+
+	pub fn datapoints_between(&self, key: &str, from: i64, to: i64, max_datapoints: usize) -> Option<Vec<(f32, i64)>> {
+		self.storage.get(key)
+			.map(|vec| {
+				let from = find_index(vec, self.base_timestamp, from);
+				let to = find_index(vec, self.base_timestamp, to);
+				let slice = &vec[from .. to];
+
+				if max_datapoints == 0 {
+					Vec::new()
+				} else if max_datapoints >= slice.len() {
+					// Just convert the slice as-is
+					slice.iter()
+						.map(|dp| dp.make_absolute(self.base_timestamp))
+						.collect()
+				} else {
+					// We have more datapoints than we need, so we need to skip some
+					(0 .. max_datapoints - 1)
```

Comment:
hmm, this looks rather complex. Why can't you just do:

```rust
slice.iter().take(max_datapoints).map(|dp| dp.make_absolute(self.base_timestamp))
```

Then, remove the if conditions?

Reply:
But we need to sample `max_datapoints` evenly from the `slice`, not just from the start.

Reply:
ok, I see but it is not evenly `shuffled` with the current implementation because it will always start with zero and then it will be increased by a multiple of `slice length` divided by `max_datapoints` together with last all the time.

Ideally, you would use [SliceRandom:shuffle](https://docs.rs/rand/0.6.5/rand/seq/trait.SliceRandom.html#tymethod.shuffle) instead and just take take the first `max_datapoints` or simply just generate random `index` where to start sample from.

Reply:
Sorry, I didn't really mean 'sample' (as-in random), I just meant that we need to get out `max_datapoints` taken from regular intervals.